File: scripts/local_python_agent.js
import http from 'http';
import { spawn } from 'child_process';
import path from 'path';
import fs from 'fs';
import os from 'os';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const PORT = 7788;

// Embedded Stego Solver Script to ensure 100% standalone execution without external dependencies
const PYTHON_STEGO_SOLVER_CODE = `import os
import sys
import re
from PIL import Image

def find_flag_in_image(image_path):
    print(f"\\n==========================================")
    print(f" [*] STANDALONE CTF SWARM PYTHON AGENT ENGINE")
    print(f" File: {image_path}")
    print(f"==========================================\\n")

    if not os.path.exists(image_path):
        print(f"[-] Error: File '{image_path}' not found.")
        return

    found_flags = set()
    flag_regex = r'(flag\\{[A-Za-z0-9_\\-]{3,80}\\}|ctf\\{[A-Za-z0-9_\\-]{3,80}\\}|ELEC\\{[A-Za-z0-9_\\-]{3,80}\\}|[a-zA-Z0-9_-]{3,15}\\{[A-Za-z0-9_\\-]{3,80}\\})'

    print("[1/4] Checking Raw Bytes & Metadata Strings...")
    with open(image_path, 'rb') as f:
        data = f.read()

    text = data.decode('latin-1', errors='ignore')
    raw_matches = re.findall(flag_regex, text, re.IGNORECASE)
    for m in raw_matches:
        if len(m) < 80 and all(32 <= ord(c) <= 126 for c in m):
            found_flags.add(m)
            print(f"  [+] Raw String Match Found: {m}")

    print("\\n[2/4] Inspecting EXIF Metadata & PNG Chunks...")
    try:
        img = Image.open(image_path)
        print(f"  Format: {img.format}, Size: {img.size}, Mode: {img.mode}")
        if img.info:
            for k, v in img.info.items():
                v_str = str(v)
                matches = re.findall(flag_regex, v_str, re.IGNORECASE)
                for m in matches:
                    if all(32 <= ord(c) <= 126 for c in m):
                        found_flags.add(m)
                        print(f"  [+] EXIF Flag Found: {m}")
    except Exception as e:
        print(f"  PIL Info: {e}")

    print("\\n[3/4] Checking Appended Data (EOF Markers)...")
    if image_path.lower().endswith('.png'):
        iend_pos = data.find(b'IEND')
        if iend_pos != -1 and iend_pos + 8 < len(data):
            extra = data[iend_pos + 8:]
            print(f"  [!] Detected {len(extra)} bytes appended after PNG IEND marker!")
            extra_text = extra.decode('latin-1', errors='ignore')
            matches = re.findall(flag_regex, extra_text, re.IGNORECASE)
            for m in matches:
                if all(32 <= ord(c) <= 126 for c in m):
                    found_flags.add(m)
                    print(f"  [+] Appended Data Flag Found: {m}")
    elif image_path.lower().endswith(('.jpg', '.jpeg')):
        eoi_pos = data.rfind(b'\\xff\\xd9')
        if eoi_pos != -1 and eoi_pos + 2 < len(data):
            extra = data[eoi_pos + 2:]
            print(f"  [!] Detected {len(extra)} bytes appended after JPEG EOI marker!")
            extra_text = extra.decode('latin-1', errors='ignore')
            matches = re.findall(flag_regex, extra_text, re.IGNORECASE)
            for m in matches:
                if all(32 <= ord(c) <= 126 for c in m):
                    found_flags.add(m)
                    print(f"  [+] Appended Data Flag Found: {m}")

    print("\\n[4/4] Running Pixel LSB Bit Plane Analysis...")
    if image_path.lower().endswith(('.png', '.bmp')):
        try:
            img = Image.open(image_path)
            if img.mode in ('RGB', 'RGBA'):
                pixels = list(img.get_flattened_data() if hasattr(img, 'get_flattened_data') else img.getdata())
                for channel_idx, channel_name in enumerate(['Red', 'Green', 'Blue']):
                    bits = [str(p[channel_idx] & 1) for p in (pixels if isinstance(pixels[0], (tuple, list)) else [pixels[i:i+3] for i in range(0, len(pixels), 3)])]
                    bit_str = ''.join(bits)
                    
                    byte_arr = bytearray()
                    for i in range(0, len(bit_str), 8):
                        b = int(bit_str[i:i+8], 2)
                        byte_arr.append(b)
                    
                    lsb_text = byte_arr.decode('latin-1', errors='ignore')
                    matches = re.findall(flag_regex, lsb_text, re.IGNORECASE)
                    for m in matches:
                        if len(m) < 80 and all(32 <= ord(c) <= 126 for c in m):
                            prefix_match = re.match(r'^([a-zA-Z0-9_-]+)\\{', m)
                            if prefix_match:
                                prefix = prefix_match.group(1).lower()
                                if len(prefix) >= 3 or prefix in ['flag', 'ctf', 'elec']:
                                    found_flags.add(m)
                                    print(f"  [+] LSB {channel_name} Channel Flag Found: {m}")
        except Exception as e:
            print(f"  LSB Warning: {e}")

    print("\\n==========================================")
    if found_flags:
        print(" [!] FINAL FLAGS DISCOVERED:")
        for flag in found_flags:
            print(f"  --> {flag}")
    else:
        print(" [i] RESULT: No Flag Found in this image")
    print("==========================================\\n")

if __name__ == '__main__':
    target = sys.argv[1] if len(sys.argv) > 1 else 'test.png'
    find_flag_in_image(target)
`;

// Locate steg_solver.py in current directory or fallback
const localScriptPath = path.join(__dirname, 'steg_solver.py');
const tempScriptPath = fs.existsSync(localScriptPath)
  ? localScriptPath
  : path.join(os.tmpdir(), 'ctf_swarm_steg_solver.py');

if (!fs.existsSync(localScriptPath) && !fs.existsSync(tempScriptPath)) {
  fs.writeFileSync(tempScriptPath, PYTHON_STEGO_SOLVER_CODE);
}

const server = http.createServer((req, res) => {
  // Allow Vercel Web app to connect
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Methods', 'GET, POST, OPTIONS');
  res.setHeader('Access-Control-Allow-Headers', 'Content-Type, Authorization');

  if (req.method === 'OPTIONS') {
    res.writeHead(204);
    res.end();
    return;
  }

  if ((req.url === '/health' || req.url === '/') && req.method === 'GET') {
    res.writeHead(200, { 'Content-Type': 'application/json' });
    res.end(JSON.stringify({
      status: 'ok',
      agent: 'CTF Swarm Desktop Agent Engine v2.0 (Stego + Recursive ZIP Solver)'
    }));
    return;
  }

  if ((req.url === '/api/analyze-stego' || req.url === '/api/analyze-archive' || req.url === '/api/analyze-file') && req.method === 'POST') {
    let bodyChunks = [];
    req.on('data', chunk => bodyChunks.push(chunk));
    req.on('end', () => {
      try {
        const body = JSON.parse(Buffer.concat(bodyChunks).toString());
        const { fileName, fileBase64, initialPassword, password, challengeText } = body;

        if (!fileBase64) {
          res.writeHead(400, { 'Content-Type': 'application/json' });
          res.end(JSON.stringify({ error: 'Missing fileBase64' }));
          return;
        }

        let chosenPwd = initialPassword || password;
        if (!chosenPwd && challengeText) {
          const m = challengeText.match(/(?:รหัส|password|pass|key|pwd)\s*[:=]?\s*([a-zA-Z0-9_\-!@#$%^&*+=~]+)/i);
          if (m) chosenPwd = m[1].trim();
        }

        const base64Data = fileBase64.replace(/^data:[^;]+;base64,/, '');
        const buffer = Buffer.from(base64Data, 'base64');
        const ext = path.extname(fileName || 'file') || (buffer.slice(0, 4).equals(Buffer.from([0x50, 0x4b, 0x03, 0x04])) ? '.zip' : '.png');
        const tempFilePath = path.join(os.tmpdir(), `ctf_upload_${Date.now()}_${path.basename(fileName || 'file', ext)}${ext}`);
        fs.writeFileSync(tempFilePath, buffer);

        // Run python with solver
        const pythonArgs = [tempScriptPath, tempFilePath];
        if (chosenPwd) {
          pythonArgs.push(chosenPwd);
        }
        const pythonProcess = spawn('python', pythonArgs);

        let stdoutData = '';
        let stderrData = '';

        pythonProcess.stdout.on('data', d => { stdoutData += d.toString(); });
        pythonProcess.stderr.on('data', d => { stderrData += d.toString(); });

        pythonProcess.on('close', code => {
          try { if (fs.existsSync(tempFilePath)) fs.unlinkSync(tempFilePath); } catch {}

          const flagRegex = /(?:flag|ctf|elec|picoctf|thm|htb|sec)[a-z0-9_-]*\{[A-Za-z0-9_\-!@#$%^&*()+=~]{3,100}\}|[a-zA-Z0-9_-]{3,15}\{[A-Za-z0-9_\-!@#$%^&*()+=~]{3,100}\}/gi;
          const matches = stdoutData.match(flagRegex) || [];

          // Clean strict 7-bit ASCII filtering
          const uniqueFlags = Array.from(new Set(matches)).filter(m => {
            for (let i = 0; i < m.length; i++) {
              if (m.charCodeAt(i) < 32 || m.charCodeAt(i) > 126) return false;
            }
            const prefixMatch = m.match(/^([a-zA-Z0-9_-]+)\{/);
            if (prefixMatch) {
              const p = prefixMatch[1].toLowerCase();
              if (p.length < 3 && !['flag', 'ctf', 'elec'].includes(p)) return false;
            }
            return true;
          });

          res.writeHead(200, { 'Content-Type': 'application/json' });
          res.end(JSON.stringify({
            success: code === 0,
            stdout: stdoutData,
            stderr: stderrData,
            flags: uniqueFlags,
            agentSource: 'Standalone User Machine Python Agent (Stego + Recursive ZIP Engine)'
          }));
        });

      } catch (err) {
        const msg = err instanceof Error ? err.message : String(err);
        res.writeHead(500, { 'Content-Type': 'application/json' });
        res.end(JSON.stringify({ error: msg }));
      }
    });
    return;
  }

  res.writeHead(404, { 'Content-Type': 'application/json' });
  res.end(JSON.stringify({ error: 'Not Found' }));
});

server.listen(PORT, () => {
  console.log(`
============================================================
  🤖 CTF SWARM STANDALONE DESKTOP AGENT (ONLINE)
============================================================
  🌐 Bridge Status : Connected & Listening on http://localhost:${PORT}
  ⚡ Engine Status : Ready to run Stego & Recursive ZIP Solvers!
  
  👉 Open https://ctfweb.vercel.app/ in your browser.
  The web app will automatically route image steganography and
  nested zip archives to run on YOUR COMPUTER'S PYTHON ENGINE!
============================================================
`);
});

File: scripts/local_python_agent.py
import http.server
import socketserver
import json
import re
import os
import sys
import tempfile
import base64
import io
from contextlib import redirect_stdout, redirect_stderr

# Ensure UTF-8 output on Windows
if sys.platform == 'win32':
    try:
        if hasattr(sys.stdout, 'reconfigure'):
            sys.stdout.reconfigure(encoding='utf-8', errors='replace')
        if hasattr(sys.stderr, 'reconfigure'):
            sys.stderr.reconfigure(encoding='utf-8', errors='replace')
    except Exception:
        pass

# Add current scripts dir to sys.path to import solvers
current_dir = os.path.dirname(os.path.abspath(__file__))
if current_dir not in sys.path:
    sys.path.insert(0, current_dir)

from archive_solver import solve_matryoshka_archive, is_archive_file, FLAG_REGEX
from steg_solver import find_flag_in_image

import socket

PORT = 7788

class CTFAgentHandler(http.server.BaseHTTPRequestHandler):
    def log_message(self, format, *args):
        pass

    def end_headers(self):
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'GET, POST, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', '*')
        self.send_header('Access-Control-Allow-Private-Network', 'true')
        super().end_headers()

    def do_OPTIONS(self):
        self.send_response(204)
        self.send_header('Content-Length', '0')
        self.end_headers()

    def do_GET(self):
        if self.path == '/health' or self.path == '/':
            resp = json.dumps({
                "status": "ok",
                "agent": "CTF Swarm Native Python Desktop Agent v2.0 (Stego + Recursive Matryoshka ZIP Engine)",
                "port": PORT
            }).encode('utf-8')
            self.send_response(200)
            self.send_header('Content-Type', 'application/json')
            self.send_header('Content-Length', str(len(resp)))
            self.send_header('Connection', 'close')
            self.end_headers()
            self.wfile.write(resp)
            return
        
        err = json.dumps({"error": "Not Found"}).encode('utf-8')
        self.send_response(404)
        self.send_header('Content-Type', 'application/json')
        self.send_header('Content-Length', str(len(err)))
        self.send_header('Connection', 'close')
        self.end_headers()
        self.wfile.write(err)

    def do_POST(self):
        if self.path in ('/api/analyze-stego', '/api/analyze-archive', '/api/analyze-file'):
            try:
                content_len = int(self.headers.get('Content-Length', 0))
                req_body = self.rfile.read(content_len).decode('utf-8')
                data = json.loads(req_body)

                file_name = data.get('fileName', 'uploaded_file')
                file_b64 = data.get('fileBase64', '')
                initial_pwd = data.get('initialPassword') or data.get('password')
                category = data.get('category', '')
                challenge_text = data.get('challengeText', '')

                # If initial_pwd not explicitly given, try extracting clues from challengeText
                if not initial_pwd and challenge_text:
                    pwd_match = re.search(r'(?:รหัส(?:ผ่าน|ชั้น[^=:\s]+)?|password(?:\s+for\s+[^=:\s]+)?|pass|key|pwd)\s*[:=]?\s*([a-zA-Z0-9_!@#$%^&*()+=~-]+)', challenge_text, re.IGNORECASE)
                    if pwd_match:
                        initial_pwd = pwd_match.group(1).strip()

                if not file_b64:
                    self.send_response(400)
                    self.send_header('Access-Control-Allow-Origin', '*')
                    self.send_header('Content-Type', 'application/json')
                    self.end_headers()
                    self.wfile.write(json.dumps({"error": "Missing fileBase64"}).encode('utf-8'))
                    return

                # Decode file
                raw_b64 = file_b64.split(',')[-1]
                file_bytes = base64.b64decode(raw_b64)

                ext = os.path.splitext(file_name)[1]
                if not ext:
                    if file_bytes.startswith(b'PK\x03\x04'):
                        ext = '.zip'
                    elif file_bytes.startswith(b'\x89PNG'):
                        ext = '.png'
                    elif file_bytes.startswith(b'\xff\xd8\xff'):
                        ext = '.jpg'

                # Save to temp file for analysis
                temp_file = tempfile.NamedTemporaryFile(delete=False, suffix=ext, prefix="ctf_agent_")
                temp_file.write(file_bytes)
                temp_file.close()

                capture_out = io.StringIO()
                capture_err = io.StringIO()
                discovered_flags = set()

                print(f"[+] Agent received task for file: {file_name} (Size: {len(file_bytes)} bytes)")

                with redirect_stdout(capture_out), redirect_stderr(capture_err):
                    if is_archive_file(temp_file.name) or category == 'archive' or file_name.lower().endswith(('.zip', '.tar', '.gz', '.bz2', '.xz', '.7z', '.rar')):
                        res = solve_matryoshka_archive(temp_file.name, initial_password=initial_pwd)
                        if res.get('flags'):
                            for f in res['flags']:
                                discovered_flags.add(f)
                    else:
                        find_flag_in_image(temp_file.name)

                try:
                    os.unlink(temp_file.name)
                except Exception:
                    pass

                stdout_str = capture_out.getvalue()
                stderr_str = capture_err.getvalue()

                # Extract any flag regex matches from stdout
                for m in FLAG_REGEX.findall(stdout_str):
                    if all(32 <= ord(c) <= 126 for c in m):
                        discovered_flags.add(m)

                resp_payload = json.dumps({
                    "success": True,
                    "stdout": stdout_str,
                    "stderr": stderr_str,
                    "flags": list(discovered_flags),
                    "agentSource": "CTF Swarm Native Python Desktop Agent"
                }).encode('utf-8')
                self.send_response(200)
                self.send_header('Content-Type', 'application/json')
                self.send_header('Content-Length', str(len(resp_payload)))
                self.send_header('Connection', 'close')
                self.end_headers()
                self.wfile.write(resp_payload)
                return

            except Exception as e:
                err_payload = json.dumps({"error": str(e)}).encode('utf-8')
                self.send_response(500)
                self.send_header('Content-Type', 'application/json')
                self.send_header('Content-Length', str(len(err_payload)))
                self.send_header('Connection', 'close')
                self.end_headers()
                self.wfile.write(err_payload)
                return

        err_resp = json.dumps({"error": "Endpoint not found"}).encode('utf-8')
        self.send_response(404)
        self.send_header('Content-Type', 'application/json')
        self.send_header('Content-Length', str(len(err_resp)))
        self.send_header('Connection', 'close')
        self.end_headers()
        self.wfile.write(err_resp)

def run_server():
    server_address = ('', PORT)
    try:
        class DualStackServer(socketserver.ThreadingMixIn, http.server.HTTPServer):
            address_family = getattr(socket, 'AF_INET6', socket.AF_INET)
            daemon_threads = True
            allow_reuse_address = True
            def server_bind(self):
                try:
                    self.socket.setsockopt(socket.IPPROTO_IPV6, socket.IPV6_V6ONLY, 0)
                except Exception: pass
                super().server_bind()
        httpd = DualStackServer(server_address, CTFAgentHandler)
    except Exception:
        class ThreadingServer(socketserver.ThreadingMixIn, http.server.HTTPServer):
            daemon_threads = True
            allow_reuse_address = True
        httpd = ThreadingServer(server_address, CTFAgentHandler)

    print(f"""
============================================================
  🤖 CTF SWARM NATIVE PYTHON DESKTOP AGENT (ONLINE)
============================================================
  🌐 Bridge Status : Connected & Listening on http://localhost:{PORT}
  ⚡ Engine Status : Ready to solve Stego & Nested Matryoshka ZIPs!
  
  👉 Open https://ctfweb.vercel.app/ (or http://localhost:5173/)
  The web app will automatically route image steganography and
  nested zip archive challenges to run on YOUR PC!
============================================================
""")
    with httpd:
        try:
            httpd.serve_forever()
        except KeyboardInterrupt:
            print("\n[!] Stopping CTF Agent Server...")

if __name__ == '__main__':
    run_server()
